python-spec/src/somacore/query/fast_csr.py

```diff
@@ -0,0 +1,299 @@
+) -> npt.NDArray[np.int64]:
+    return cast(
+        npt.NDArray[np.int64], index.get_indexer(ids).astype(target_dtype, copy=False)
+    )
+
+
+def _read_csr(
+    matrix: scd.SparseNDArray, obs_joinids: pa.Array, var_joinids: pa.Array
+) -> Tuple[
+    npt.NDArray[np.number],  # data
+    npt.NDArray[np.integer],  # indptr
+    npt.NDArray[np.integer],  # indices
+    Tuple[int, int],  # shape
+]:
+    if not isinstance(matrix, scd.SparseNDArray) or matrix.ndim != 2:
+        raise TypeError("read_scipy_csr can only read from a 2D SparseNDArray")
+
+    max_workers = (os.cpu_count() or 4) + 2
+    with futures.ThreadPoolExecutor(max_workers=max_workers) as pool:
+        acc = _CSRAccumulator(
+            obs_joinids=obs_joinids, var_joinids=var_joinids, pool=pool
+        )
+        for tbl in eager_iter._EagerIterator(
+            matrix.read((obs_joinids, var_joinids)).tables(),
+            pool=pool,
+        ):
+            acc.append(tbl["soma_data"], tbl["soma_dim_0"], tbl["soma_dim_1"])
+
+        data, indptr, indices, shape = acc.finalize()
+
+    return data, indptr, indices, shape
+
+
+def read_scipy_csr(
+    matrix: scd.SparseNDArray, obs_joinids: pa.Array, var_joinids: pa.Array
+) -> sparse.csr_matrix:
+    """
+    Given a 2D SparseNDArray and joinids for the two dimensions, read the
+    slice and return it as an SciPy sparse.csr_matrix.
+    """
+    data, indptr, indices, shape = _read_csr(matrix, obs_joinids, var_joinids)
+    csr = _create_scipy_csr_matrix(data, indices, indptr, shape=shape)
+    return csr
+
+
+def read_arrow_csr(
+    matrix: scd.SparseNDArray, obs_joinids: pa.Array, var_joinids: pa.Array
+) -> pa.SparseCSRMatrix:
+    """
+    Given a 2D SparseNDArray and joinids for the two dimensions, read the
+    slice and return it as an SciPy sparse.csr_matrix.
+    """
+    data, indptr, indices, shape = _read_csr(matrix, obs_joinids, var_joinids)
+    csr = pa.SparseCSRMatrix.from_numpy(data, indptr, indices, shape=shape)
+    return csr
+
+
+def _create_scipy_csr_matrix(
+    data: npt.NDArray[np.number],
+    indices: npt.NDArray[np.integer],
+    indptr: npt.NDArray[np.integer],
+    shape: Tuple[int, int],
+) -> sparse.csr_matrix:
+    """
+    Create a Scipy sparse.csr_matrix from component elements. Conceptually this
+    is identical to:
+        sparse.csr_matrix((data, indices, indptr), shape=shape)
+
+    This ugliness is to bypass the O(N) scan that scipy.sparse._cs_matrix.__init__
+    does when a new compressed matrix is created.
+
+    See https://github.com/scipy/scipy/issues/11496 for details on the bug.
+    """
+    matrix = sparse.csr_matrix.__new__(sparse.csr_matrix)
+    matrix.data = data
+    matrix.indices = indices
+    matrix.indptr = indptr
+    matrix._shape = shape
+    return matrix
+
+
+def _SparseCSRMatrix_to_scipy(csr: pa.SparseCSRMatrix) -> sparse.csr_matrix:
+    """
+    Convert Arrow SparseCSRMatrix to scipy.sparse.csr_matrix. Semantically the same
```

Comment:
Convert pyarrow.SparseCSRMatrix to scipy.sparse.csr_matrix ...

Reply:
pyarrow is simply a wrapper on Arrow.